pkg/chunk/cache/memcached.go

```diff
@@ -86,6 +109,64 @@ func (c *Memcached) Fetch(ctx context.Context, keys []string) (found []string, b
 	return
 }
 
+func (c *Memcached) fetchKeysBatched(ctx context.Context, keys []string) (found []string, bufs [][]byte, missed []string, err error) {
+	type result struct {
+		found  []string
+		bufs   [][]byte
+		missed []string
+		err    error
+	}
+
+	inputCh := make(chan []string)
+	resultsCh := make(chan *result)
+	var wg sync.WaitGroup
+
+	// Initialise workers.
+	for i := 0; i < c.cfg.Parallelism; i++ {
+		wg.Add(1)
+		go func(inputCh <-chan []string, resultsCh chan<- *result) {
+			for keys := range inputCh {
+				found, bufs, missed, err := c.Fetch(ctx, keys)
+				resultsCh <- &result{
+					found:  found,
+					bufs:   bufs,
+					missed: missed,
+					err:    err,
+				}
+			}
+
+			wg.Done()
+		}(inputCh, resultsCh)
+	}
+
+	go func() {
+		batchSize := c.cfg.BatchSize
+		for i := 0; i < len(keys); i += batchSize {
+			batchKeys := keys[i:util.Min(i+batchSize, len(keys))]
+			inputCh <- batchKeys
+		}
+
+		close(inputCh)
+
+		// Wait for all in-flight requests to finish.
+		wg.Wait()
+		close(resultsCh)
+	}()
+
+	for result := range resultsCh {
+		if result.err != nil {
+			err = result.err
+			return
```

Comment:
Exiting early like this will leave goroutines blocked on sends to resultCh; this will hold them live and prevent them being GC'd.  Either loop through all the results and return the last error (prefferred) or buffer the channel.